Sources/MessagingInApp/Gist/Managers/MessageManager.swift

```diff
@@ -50,23 +52,40 @@ class MessageManager: EngineWebDelegate {
         subscribeToInAppMessageState()
     }
 
+    deinit {
+        unsubscribeFromInAppMessageState()
+        removeEngineWebView()
+    }
+
     func subscribeToInAppMessageState() {
         // Keep a strong reference to the subscriber to prevent deallocation and continue receiving updates
         // Also, since we do not store strong reference of MessageManager anywhere, not keeping a strong reference of
         // subscriber may result in deallocation of MessageManager and hence dismissal of message unexpectedly.
         inAppMessageStoreSubscriber = {
             let subscriber = InAppMessageStoreSubscriber { [self] state in
-                switch state.currentMessageState {
+                let messageState = state.currentMessageState
+                switch messageState {
                 case .displayed:
                     threadUtil.runMain {
                         self.loadModalMessage()
                     }
 
-                case .dismissed:
+                // Dismiss the message when the message is dismissed or initial state
+                // Initial state may only be received when state is reset while a message was being displayed
+                case .dismissed, .initial:
                     threadUtil.runMain {
-                        self.engine.delegate = nil
-                        self.dismissMessage()
-                        self.inAppMessageStoreSubscriber = nil
+                        self.removeEngineWebView()
+                        // Unsubscribe from InAppMessageState when the message is dismissed completely
+                        // so that MessageManager is deallocated only after dismiss animation is completed.
+                        self.dismissMessage {
+                            self.unsubscribeFromInAppMessageState()
+                            // Fetch user messages from local store after message is dismissed so that
+                            // next message can be displayed instantly if available.
+                            // This is only needed when message is dismissed and not when it is reset.
+                            if case .dismissed = messageState {
+                                self.gist.fetchUserMessagesFromStoreState()
```

Comment:
In Android, this is done by fetching messages from remote in `onActivityResumed` in `GistSdk`

Reply:
I wonder whats stopping us from having this as a middleware, that any time `dismiss` action occurs we process the current local queue.

Reply:
Discussed this on the call. Since the middleware doesn't account for message animation, an additional `loading` event is received with suggested approach on current `MessageManager` while previous message is being dismissed. To avoid this event and make the behavior more consistent with Android, we followed similar approach by fetching messages from remote queue instead.